pkg/policy/api/cidr.go

```diff
@@ -42,3 +48,41 @@ type CIDRRule struct {
 	// or provided by user
 	Generated bool `json:"-"`
 }
+
+// CIDRSlice is a slice of CIDRs. It allows receiver methods to be defined for
+// transforming the slice into other convenient forms such as
+// EndpointSelectorSlice.
+type CIDRSlice []CIDR
+
+// CIDRRuleSlice is a slice of CIDRRules. It allows receiver methods to be
+// defined for transforming the slice into other convenient forms such as
+// EndpointSelectorSlice.
+type CIDRRuleSlice []CIDRRule
+
+// ComputeResultantCIDRSet converts a slice of CIDRRules into a slice of
+// individual CIDRs. This expands the cidr defined by each CIDRRule, applies
+// the CIDR exceptions defined in "ExceptCIDRs", and forms a minimal set of
+// CIDRs that cover all of the CIDRRules.
+func ComputeResultantCIDRSet(cidrs CIDRRuleSlice) CIDRSlice {
+	var allResultantAllowedCIDRs CIDRSlice
+	for _, s := range cidrs {
+		// No need for error checking, as CIDRRule.Sanitize() already does.
+		_, allowNet, _ := net.ParseCIDR(string(s.Cidr))
+
+		var removeSubnets []*net.IPNet
+		for _, t := range s.ExceptCIDRs {
+			// No need for error checking, as CIDRRule.Sanitize() already
+			// does.
+			_, removeSubnet, _ := net.ParseCIDR(string(t))
+			removeSubnets = append(removeSubnets, removeSubnet)
+		}
+		// No need for error checking, as have already validated that none of
+		// the possible error cases can occur in ip.RemoveCIDRs
+		resultantAllowedCIDRs, _ := ip.RemoveCIDRs([]*net.IPNet{allowNet}, removeSubnets)
```

Comment:
Do we also want to call `ip.CoalesceCIDRs` ?
Was the rational to add this function and types here because we need the auto-generated serialisation functions? It seems like a useful extension to the other package.

Reply:
The rationale here is that #3835 will add some additional helper functions for CIDR / labels conversion, which will need this function. I didn't attempt to change any logic.

https://github.com/cilium/cilium/issues/1658 tracks making use of `ip.CoalesceCIDRs`.